### PR TITLE
Extend bulk benchmark

### PR DESCRIPTION
**Pull request by Matthias247**
This change improves the bulk benchmark:
- The total amount of requests the benchmark performs can be configured
- The amount of concurrent requests on a given connection can be
  configured
- The amount of data to transfer per stream can be configured

E.g.

```
bulk --streams 300 --max_streams 100 --stream_size 10
```

will create 300 streams - with a maximum of 100 active streams at a time,
and send 10MB of data on each stream.

The "short" cli flags are taken form h2load where applicable.

The benchmark now also shows metrics which indicate the performance
of individual streams, which allows to judge fairness. Example output:

```
Overall stats:

Sent 3145728000 bytes on 300 streams in 12.11s (247.81 MiB/s)

Stream metrics:

      │  Throughput   │ Duration
──────┼───────────────┼──────────
 AVG  │  149.17 MiB/s │  396.00ms
 P0   │    0.83 MiB/s │   39.00ms
 P10  │   51.19 MiB/s │   40.00ms
 P50  │  129.75 MiB/s │   77.00ms
 P90  │  247.00 MiB/s │  195.00ms
 P100 │  254.75 MiB/s │    12.11s
```

### CODE REVIEW

File: bench/src/bulk.rs [hunk 1]
```diff
@@ -51,27 +62,72 @@ async fn server(mut incoming: quinn::Incoming) -> Result<()> {
...
+    loop {
+        let mut stream = match uni_streams.next().await {
+            None => return Ok(()),
+            Some(Err(quinn::ConnectionError::ApplicationClosed(_))) => return Ok(()),
+            Some(Err(e)) => return Err(e).context("accepting stream failed"),
+            Some(Ok(stream)) => stream,
+        };
+        trace!("stream established");
+
+        let _jh: tokio::task::JoinHandle<Result<()>> = tokio::spawn(async move {
```
Comment (djc) [inline bench/src/bulk.rs:75]: Why do we need to name the type of the `JoinHandle` here?

File: bench/src/bulk.rs [hunk 2]
```diff
@@ -51,27 +62,72 @@ async fn server(mut incoming: quinn::Incoming) -> Result<()> {
...
+}
+
+#[derive(Debug)]
+struct SendResult {
+    duration: Duration,
+    size: usize,
+    throughput: f64,
+}
+
+async fn send_data_on_stream(
```
Comment (djc) [inline bench/src/bulk.rs:93]: Can we split out a separate commit that extracts `send_data_on_stream()` out of `client()`?

File: bench/src/bulk.rs [hunk 3]
```diff
@@ -51,27 +62,72 @@ async fn server(mut incoming: quinn::Incoming) -> Result<()> {
...
+            while let Some((data, offset)) = stream.read_unordered().await? {
+                n = n.max(offset + data.len() as u64);
+            }
+
+            Ok(())
+        });
+    }
+}
+
+#[derive(Debug)]
```
Comment (djc) [inline bench/src/bulk.rs:86]: Since `SendResult` is used by `send_data_on_stream()`, let's move the type below the function defn.

File: bench/src/bulk.rs [hunk 4]
```diff
@@ -80,34 +136,143 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
...
+    let connection = Arc::new(connection);
+    let get_connection = || connection.clone();
+
+    let mut ops = futures::stream::iter((0..config.streams).map(|_| async move {
+        send_data_on_stream(get_connection(), config.stream_size_mb).await
+    }))
+    .buffer_unordered(config.max_streams);
+
+    let mut total_size = 0;
+    let mut duration_hist: Histogram<u64> = Histogram::new(3).unwrap();
```
Comment (djc) [inline bench/src/bulk.rs:160]: Can we remove the need for the type annotations here by adding a `u64` suffix to the `new()` parameter?

File: bench/src/bulk.rs [hunk 5]
```diff
@@ -80,34 +136,143 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
...
+
 fn rt() -> Runtime {
     Builder::new()
         .basic_scheduler()
         .enable_all()
         .build()
         .unwrap()
 }
+
+fn transport_config() -> quinn::TransportConfig {
```
Comment (djc) [inline bench/src/bulk.rs:238]: I think I'd prefer to keep this inline and just clone the `transport_config`, what do you think?

File: bench/src/bulk.rs [hunk 6]
```diff
@@ -80,34 +136,143 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
...
+
+    println!("Stream metrics:\n");
+
+    println!(
+        "AVG : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
+        throughput_hist.mean() as f64 / 1024.0 / 1024.0,
+        Duration::from_millis(duration_hist.mean() as u64)
+    );
+    println!(
+        "P0  : Throughput: {:7.2} MiB/s, Duration: {:.2?}",
```
Comment (djc) [inline bench/src/bulk.rs:192]: I find the ordering a bit counterintuitive here, I guess because percentiles are more usually used for latencies, so I expect the highest percentile to be worst and the lowest percentile to be best. You probably have more experience with throughput distribution representations than me, though...

**Review by djc — COMMENTED** [review]
This looks very useful! Some questions/suggestions about style choices follow...

File: bench/src/bulk.rs (hunk 1 above)
Comment (Matthias247) [reply bench/src/bulk.rs]: We have to explicitely write the type so that `?` works. `_jh` can be replaced with `_`. I typically like to still see what gets ignored based on the variable name. But it's obviously shown here by the type parameter. I can remove it.

File: bench/src/bulk.rs (hunk 3 above)
Comment (Matthias247) [reply bench/src/bulk.rs]: That's the reason why I have it above: I like to learn about dependencies before the places where they are used. Allows to read from top to bottom :-) 

But I will change it if preferred

File: bench/src/bulk.rs (hunk 2 above)
Comment (Matthias247) [reply bench/src/bulk.rs:86]: We could. But given that half of the changes are still separate (e.g. measuring durations and having a return value which contains metrics), I don't think this is a super worthwhile exercise.

File: bench/src/bulk.rs (hunk 4 above)
Comment (Matthias247) [reply bench/src/bulk.rs]: done

File: bench/src/bulk.rs (hunk 5 above)
Comment (Matthias247) [reply bench/src/bulk.rs:238]: It's created 2 times - once for the server and once for the client in 2 different functions. I think when trying to play around with new values it's easiest to modify both in one place. Therefore I moved it out (and actually there wasn't any explicit configuration before).

File: bench/src/bulk.rs (hunk 6 above)
Comment (Matthias247) [reply bench/src/bulk.rs:192]: I don't think there is a right or wrong here. Both approaches work. I actually want to add first-byte latency as another column, and there P0 is the best value and P100 the worst 🤷‍♂️

File: bench/src/bulk.rs (hunk 3 above)
Comment (Ralith) [reply bench/src/bulk.rs]: I tend to come down on @djc's side here as well, as it's easy to overflow your working memory with dependencies before you get to the root, and the root can be a useful guide to understand the significance of the dependencies.

File: bench/src/bulk.rs (hunk 6 above)
Comment (Ralith) [reply bench/src/bulk.rs]: I was confused about the example output because it seemed to suggest that the lowest-throughput stream had the shortest duration. Not sure how to do better without making it massively less compact, though.

Comment (Matthias247) [reply bench/src/bulk.rs]: I  agree. If one interprets it as "these are the values of individual streams", it gets a bit confusing. It should be clear that those are not, because we have far more streams, but 🤷‍♂️

One thing that can be changed is to make it a real table, and move the throughput / duration labels to a separate row. Not sure if this is clearer or not however.

Individual stream metrics are found in the the trace:

```
Dec 27 12:00:39.056  INFO bulk: stream finished: Ok(SendResult { duration: 11.809008s, size: 10485760, throughput: 887945.8799587569 })
```

Comment (Matthias247) [reply bench/src/bulk.rs]: Here's the table format. I think I slightly prefer it:

```
Stream metrics:

      │  Throughput   │ Duration
──────┼───────────────┼──────────
 AVG  │  155.43 MiB/s │  391.00ms
 P0   │    0.84 MiB/s │   39.00ms
 P10  │   50.56 MiB/s │   40.00ms
 P50  │  131.25 MiB/s │   76.00ms
 P90  │  249.87 MiB/s │  195.00ms
 P100 │  255.25 MiB/s │    11.97s
```

File: bench/src/bulk.rs [hunk 7]
```diff
@@ -80,34 +136,128 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
...
+
+#[derive(Debug, Clone, Copy)]
+struct BenchConfig {
+    /// The total number of streams which should be created
+    streams: usize,
+    /// The amount of concurrent streams which should be used
+    max_streams: usize,
+    /// The amount of data to transfer on a stream in megabytes
+    stream_size_mb: usize,
+}
```
Comment (Ralith) [inline bench/src/bulk.rs:253]: Can we really not just derive `Clone + Copy` on `Opt`?

File: bench/src/bulk.rs [hunk 8]
```diff
@@ -51,27 +62,72 @@ async fn server(mut incoming: quinn::Incoming) -> Result<()> {
...
+            Some(Err(e)) => return Err(e).context("accepting stream failed"),
+            Some(Ok(stream)) => stream,
+        };
+        trace!("stream established");
+
+        let _: tokio::task::JoinHandle<Result<()>> = tokio::spawn(async move {
+            let mut n = 0u64;
+            while let Some((data, offset)) = stream.read_unordered().await? {
+                n = n.max(offset + data.len() as u64);
+            }
```
Comment (Ralith) [inline bench/src/bulk.rs:79]: What are we keeping track of `n` for?

File: bench/src/bulk.rs [hunk 9]
```diff
@@ -80,34 +136,128 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
...
-    const DATA: &[u8] = &[0xAB; 1024 * 1024];
     let start = Instant::now();
-    for _ in 0..1024 {
-        stream
-            .write_all(DATA)
-            .await
-            .context("failed sending data")?;
+
+    let connection = Arc::new(connection);
+    let get_connection = || connection.clone();
```
Comment (Ralith) [inline bench/src/bulk.rs:152]: Style nit: this feels really cheesy. Let's just move a clone into the closure, even if that means doing a dance with blocks or naming something `connection2`.

**Review by Ralith — COMMENTED** [review]
Great improvements overall!

File: bench/src/bulk.rs (hunk 7 above)
Comment (Matthias247) [reply bench/src/bulk.rs:253]: Sure. I tried to separate the CLI options and function arguments. But for such a small program it likely doesn't matter.

File: bench/src/bulk.rs (hunk 8 above)
Comment (Matthias247) [reply bench/src/bulk.rs:79]: leftover from the old version. Will remove

File: bench/src/bulk.rs (hunk 9 above)
Comment (Matthias247) [reply bench/src/bulk.rs:152]: I don't get what is recommended here. The workaround with the closure is required to get access to the connection in each stream map iteration (which happens only when the whole future is awaited). A single clone won't be sufficient.

File: bench/src/bulk.rs (hunk 5 above)
Comment (djc) [reply bench/src/bulk.rs:219]: Sorry, must have missed that -- I did find the two usages but thought they were in the same function.

File: bench/src/bulk.rs (hunk 6 above)
Comment (djc) [reply bench/src/bulk.rs]: Yeah, I like the table format, too.

File: bench/src/bulk.rs [hunk 10]
```diff
@@ -80,34 +128,112 @@ async fn client(server_addr: SocketAddr, server_cert: quinn::Certificate) -> Res
...
+
+    let mut ops =
+        futures::stream::iter((0..opt.streams).map(|_| async move {
+            send_data_on_stream(get_connection(), opt.stream_size_mb).await
+        }))
```
Comment (djc) [inline bench/src/bulk.rs:149]: Something like this?

```suggestion
        futures::stream::iter((0..opt.streams).map(|_| {
            let connection = connection.clone();
            async move {
                send_data_on_stream(connection, opt.stream_size_mb).await
            }
        }))
```

File: bench/src/bulk.rs (hunk 1 above)
Comment (djc) [reply bench/src/bulk.rs]: Okay. Maybe also just import `tokio::task::JoinHandle` up top so that we don't have the whole path inline here.

File: bench/src/bulk.rs (hunk 2 above)
Comment (djc) [reply bench/src/bulk.rs:82]: I do feel that it makes changes like this substantially easier to review, which I feel is valuable even though splitting the commit (or redoing the changes) can feel like drudgery. So while I understand your reluctance, I would appreciate it.

File: bench/src/bulk.rs (hunk 10 above)
Comment (Matthias247) [reply bench/src/bulk.rs]: I don't think that is any improvement (as a reader now the first thing I'm asking myself is what this weird extra block is doing there), but I still applied it.

Frankly speaking: I feel there is a bit of too much bikeshedding going on. This comments all lead to spending effort to move code back and forth without any objective reason why it should be in one particular place, and without any functional changes. If this is highly important to you, please consider writing a styleguide

Comment (djc) [reply bench/src/bulk.rs]: That's fair, I'll see if I can write a style guide soonish. On the other hand, I'm not sure it would have covered everything we commented on here. In my experience, we routinely we give each other feedback on the clearest way to express something in code/submit something for review, and make changes accordingly, and it generally has been fine.

**Review by djc — APPROVED** [review]
